src/python/grpcio_testing/grpc_testing/_channel.py

```diff
@@ -0,0 +1,753 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Channel fixtures."""
+
+import collections
+import logging
+import threading
+
+import grpc
+import grpc_testing
+from grpc_testing import _common
+
+_NOT_YET_OBSERVED = object()
+
+
+class _RpcState(object):
+
+    def __init__(self, invocation_metadata, requests, requests_closed):
+        self.condition = threading.Condition()
+        self.invocation_metadata = invocation_metadata
+        self.requests = requests
+        self.requests_closed = requests_closed
+        self.initial_metadata = None
+        self.responses = []
+        self.trailing_metadata = None
+        self.code = None
+        self.details = None
+
+
+def _state_add_request(state, request):
+    with state.condition:
+        if state.code is None and not state.requests_closed:
+            state.requests.append(request)
+            state.condition.notify_all()
+            return True
+        else:
+            return False
+
+
+def _state_no_more_requests(state):
+    with state.condition:
+        if state.code is None and not state.requests_closed:
+            state.requests_closed = True
+            state.condition.notify_all()
+
+
+def _state_take_response(state):
+    with state.condition:
+        while True:
+            if state.code is grpc.StatusCode.OK:
+                if state.responses:
+                    response = state.responses.pop(0)
+                    return _common.ChannelRpcRead(response, None, None, None)
+                else:
+                    return _common.ChannelRpcRead(None, state.trailing_metadata,
+                                                  grpc.StatusCode.OK,
+                                                  state.details)
+            elif state.code is None:
+                if state.responses:
+                    response = state.responses.pop(0)
+                    return _common.ChannelRpcRead(response, None, None, None)
+                else:
+                    state.condition.wait()
+            else:
+                return _common.ChannelRpcRead(None, state.trailing_metadata,
+                                              state.code, state.details)
+
+
+def _state_cancel(state, code, details):
+    with state.condition:
+        if state.code is None:
+            if state.initial_metadata is None:
+                state.initial_metadata = _common.FUSSED_EMPTY_METADATA
+            state.trailing_metadata = _common.FUSSED_EMPTY_METADATA
+            state.code = code
+            state.details = details
+            state.condition.notify_all()
+            return True
+        else:
+            return False
+
+
+def _state_terminate(state):
+    with state.condition:
+        while True:
```

Comment:
I remain unconvinced. I don't have much disagreement there, though I would question the "clearly better" part even there, but I think we might be arguing a straw-man here, since the actual code in question is much simpler with no branches within the loop (by the way, the two code snippets are actually not semantically equivalent (due to the presence of the `with` clause), but that's besides the point).

I think my critical question is under what circumstances would we ever want to prefer `while condition:` over `while True:`? It's hard for me to imagine a `while` loop getting much simpler than @griffithjames's version.

It would remain an acceptable and internally consistent conclusion if we agree that `while condition` is a universally bad idea; do you think that's the case though? If not, what's the principle that is preventing us from using `while True:` everywhere?